src/log.rs

```diff
@@ -1,47 +1,152 @@
+        // or the Rust API can be used instead.
+        let fmt = cstr!("%c");
+        for &byte in s.as_bytes() {
+            // # Safety
+            //
+            // The first argument is `self.cookie`
+            // and the rest are safe to call `printf` with,
+            // as required by [`Self::new`].
+            unsafe {
+                (self.callback)(self.cookie, fmt.as_ptr(), byte as c_uint);
+            }
         }
+        Ok(())
     }
 }
 
-#[cold]
-pub unsafe extern "C" fn rav1d_log(c: *mut Rav1dContext, format: *const c_char, args: ...) {
-    if validate_input!(!c.is_null()).is_err() {
-        return;
+#[derive(Clone, Default)]
+pub(crate) enum Rav1dLogger {
+    Dav1d(OnlyDav1dLogger),
+    Stdout,
+    #[default]
+    Stderr,
+}
+
+/// Any type implementing [`Rav1dLog`] can be used with [`write!`].
+pub trait Rav1dLog {
+    fn write_fmt(&self, args: fmt::Arguments);
+}
+
+impl Rav1dLog for Rav1dLogger {
+    /// Logging doesn't have to be fast when it's on,
+    /// but we don't want it slow things down when it's off,
+    /// so ensure the logging code isn't inlined everywhere, bloating call sites.
+    #[inline(never)]
+    fn write_fmt(&self, args: fmt::Arguments) {
+        match self {
+            // The `dav1d.clone()` is because [`fmt::Write::write_fmt`] takes `&mut`
+            // even though we don't need it to.
+            // [`OnlyDav1dLogger`] is trivial to [`Clone`], though, so we can just do that.
+            Self::Dav1d(dav1d) => dav1d.clone().write_fmt(args).unwrap(),
+            Self::Stdout => stdout().write_fmt(args).unwrap(),
+            Self::Stderr => stderr().write_fmt(args).unwrap(),
+        }
     }
-    if ((*c).logger.callback).is_none() {
-        return;
+}
+
+impl Rav1dLog for Option<Rav1dLogger> {
+    /// When a logger isn't set, we don't want to have to run any code here,
+    /// so force this to be inlined so a [`None`] can be seen.
+    #[inline(always)]
+    fn write_fmt(&self, args: fmt::Arguments) {
+        if let Some(logger) = self {
+            logger.write_fmt(args);
+        }
+    }
+}
+
+/// Used as a marker for [`Rav1dLogger::Stdout`].  Still a valid (i.e. safe) [`Dav1dLoggerCallback`], though.
+unsafe extern "C" fn rav1d_logger_stdout(_cookie: *mut c_void, _fmt: *const c_char, ...) {}
+
+/// Used as a marker for [`Rav1dLogger::Stderr`].  Still a valid (i.e. safe) [`Dav1dLoggerCallback`], though.
+unsafe extern "C" fn rav1d_logger_stderr(_cookie: *mut c_void, _fmt: *const c_char, ...) {}
+
+impl From<Dav1dLogger> for Option<Rav1dLogger> {
+    fn from(logger: Dav1dLogger) -> Self {
+        let Dav1dLogger { cookie, callback } = logger;
+        let callback = callback?;
+        Some(if callback == rav1d_logger_stdout {
+            Rav1dLogger::Stdout
+        } else if callback == rav1d_logger_stderr {
+            Rav1dLogger::Stderr
+        } else {
+            Rav1dLogger::Dav1d(OnlyDav1dLogger { cookie, callback })
+        })
```

Comment:
Match?

Reply:
You can't match on function pointers.

Reply:
Meant something like this. Minor thing, just a style nitpick, but that if chain was a bit less clear than it could be.

```rust
match callback {
    Some(f) if f == rav1d_logger_stdout => Rav1dLogger::Stdout,
    Some(f) ...
    Some(f) => Rav1dLogger::Dav1d(...),
    None => None,
}
```

Reply:
Oh sure I can do that.  People have said they don't like that in the past but it does look clean.

Reply:
Yeah, this is definitely just a style nitpick, I think it would be clean in this case but it's up to you.

Reply:
Changed in b368117.